### PR TITLE
feat(esl-utils): extend SynteticEventTarget with ability to separate events enhancement

### PR DESCRIPTION
**Pull request by fshovchko**
In the scope of this PR I extended SynteticEventTarget to separate events by their type.
The end result repeats implementation of native event storage.
Code base tests were extended and refactored.

### CODE REVIEW

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-utils/dom/events/target.ts [hunk 1]
```diff
@@ -15,29 +19,42 @@ export class SyntheticEventTarget implements EventTarget {
...
 
   public addEventListener(callback: EventListenerOrEventListenerObject): void;
-  public addEventListener(type: 'change', callback: EventListenerOrEventListenerObject): void;
-  public addEventListener(type: any, callback: EventListenerOrEventListenerObject = type): void {
+  public addEventListener(type: string, callback: EventListenerOrEventListenerObject): void;
+  public addEventListener(type: string | EventListenerOrEventListenerObject = 'change', callback?: EventListenerOrEventListenerObject): void {
+    if (typeof type !== 'string') {
+      callback = type;
+      type = 'change';
+    }
```
Comment (ala-n) [inline src/modules/esl-utils/dom/events/target.ts:31]:
```suggestion
    if (typeof type !== 'string') return this,addEventListener('change', type);
```

Comment (ala-n) [reply src/modules/esl-utils/dom/events/target.ts:31]: + the same trick in the `hasEventListener` method

File: src/modules/esl-utils/dom/events/target.ts [hunk 2]
```diff
@@ -15,29 +19,42 @@ export class SyntheticEventTarget implements EventTarget {
...
     }
-    return this._listeners.size > minCount;
+    return this._listeners[type]?.length > minCount;
   }
 
   public addEventListener(callback: EventListenerOrEventListenerObject): void;
-  public addEventListener(type: 'change', callback: EventListenerOrEventListenerObject): void;
-  public addEventListener(type: any, callback: EventListenerOrEventListenerObject = type): void {
+  public addEventListener(type: string, callback: EventListenerOrEventListenerObject): void;
+  public addEventListener(type: string | EventListenerOrEventListenerObject = 'change', callback?: EventListenerOrEventListenerObject): void {
```
Comment (ala-n) [inline src/modules/esl-utils/dom/events/target.ts:27]:
```suggestion
  public addEventListener(type: string | EventListenerOrEventListenerObject, callback?: EventListenerOrEventListenerObject): void {
```
Method does not have empty argument definition

File: src/modules/esl-utils/dom/events/target.ts [hunk 3]
```diff
@@ -15,29 +19,42 @@ export class SyntheticEventTarget implements EventTarget {
...
 
   public removeEventListener(callback: EventListenerOrEventListenerObject): void;
-  public removeEventListener(type: 'change', callback: EventListenerOrEventListenerObject): void;
-  public removeEventListener(type: any, callback: EventListenerOrEventListenerObject = type): void {
+  public removeEventListener(type: string, callback: EventListenerOrEventListenerObject): void;
+  public removeEventListener(type: string | EventListenerOrEventListenerObject = 'change', callback?: EventListenerOrEventListenerObject): void {
+    if (typeof type !== 'string') {
+      callback = type;
+      type = 'change';
+    }
```
Comment (ala-n) [inline src/modules/esl-utils/dom/events/target.ts:44]:
```suggestion
  public removeEventListener(type: string | EventListenerOrEventListenerObject, callback?: EventListenerOrEventListenerObject): void {
    if (typeof type !== 'string') return this.removeEventListener('change', type);
```

File: src/modules/esl-utils/dom/events/target.ts [hunk 4]
```diff
@@ -15,29 +19,42 @@ export class SyntheticEventTarget implements EventTarget {
...
+  public addEventListener(type: string | EventListenerOrEventListenerObject = 'change', callback?: EventListenerOrEventListenerObject): void {
+    if (typeof type !== 'string') {
+      callback = type;
+      type = 'change';
+    }
+
     validateEventListenerType(callback);
-    this._listeners.add(callback);
+    const listeners = this._listeners[type];
+    this.hasEventListener(type) ? listeners.push(callback!) : Object.assign(this._listeners, {[type]: [callback!]});
```
Comment (ala-n) [inline src/modules/esl-utils/dom/events/target.ts:35]:
```suggestion
    if (this._listeners[type] && !this._listeners[type].includes(callback!)) this._listeners[type].push(callback!)
    else Object.assign(this._listeners, {[type]: [callback!]});
```

File: src/modules/esl-utils/dom/events/target.ts [hunk 5]
```diff
@@ -15,29 +19,42 @@ export class SyntheticEventTarget implements EventTarget {
...
+    if (typeof type !== 'string') {
+      callback = type;
+      type = 'change';
+    }
+
     validateEventListenerType(callback);
-    this._listeners.delete(callback);
+    const listeners = this._listeners[type];
+    const cbIndex = listeners?.indexOf(callback!);
+    if (cbIndex > -1) listeners.splice(cbIndex, 1);
```
Comment (ala-n) [inline src/modules/esl-utils/dom/events/target.ts:49]: Cosmetic: 
```suggestion
    if (!this._listeners[type]) return;
    this._listeners[type] = this._listeners[type].filter((cb) => cb !== callback)
```

**Review by ala-n — CHANGES_REQUESTED** [review]

File: src/modules/esl-utils/dom/events/target.ts (hunk 4 above)
Comment (ala-n) [reply src/modules/esl-utils/dom/events/target.ts:24]: The current implementation accepts a single listener to be subscribed multiple times, but it's unacceptable as this fact may used by the consumers.

**Review by julia-murashko — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

**Review by yadamskaya — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

**Review by ala-n — APPROVED** [review]

**Pull request merged**